pkg/controller/metrics/metrics.go

```diff
@@ -818,6 +818,54 @@ func readMetricsConfig() (*metricsconfig.MetricsConfig, error) {
 	return config, nil
 }
 
+// GetOptionalClusterTypeLabels reads the MetricsWithClusterTypeLabels from the metrics config and returns the same as a
+// map, as well as a slice of metric labels. The latter is helpful for defining the relevant metrics
+func GetOptionalClusterTypeLabels(mConfig *metricsconfig.MetricsConfig) (map[string]string, []string) {
+	var mapClusterTypeLabelToValue = make(map[string]string)
+	var optionalClusterTypeLabels []string
+	for k, v := range mConfig.MetricsWithClusterTypeLabels {
+		mapClusterTypeLabelToValue[k] = v
+		optionalClusterTypeLabels = append(optionalClusterTypeLabels, k)
+	}
+	return mapClusterTypeLabelToValue, optionalClusterTypeLabels
+}
+
+// LogHistogramMetricWithOptionalLabels is used to log a histogram type metric with support for optional cluster type
+// labels. Caller should ensure there isn't an overlap between fixed and optional labels.
+func LogHistogramMetricWithOptionalLabels(metric *prometheus.HistogramVec, value float64, obj metav1.Object,
+	fixedLabels map[string]string, optionalLabels map[string]string, log log.FieldLogger) {
+	// Build a partial label vector for non-optional metrics
+	curriedVec, err := metric.CurryWith(fixedLabels)
+	// Add optional labels, if present. Labels added here must not have been already added in the step above
+	for k, v := range optionalLabels {
+		curriedVec, err = curriedVec.CurryWith(prometheus.Labels{k: GetClusterDeploymentType(obj, v)})
+	}
+	if err != nil {
+		// The possibility of hitting an error here is low, but report it and don't observe the metric in case it happens
+		log.WithError(err).Error("error observing metric")
+	} else {
+		curriedVec.WithLabelValues().Observe(value)
+	}
+}
+
+// LogCounterMetricWithOptionalLabels is used to log a counter type metric with support for optional cluster type
+// labels. Caller should ensure there isn't an overlap between fixed and optional labels.
+func LogCounterMetricWithOptionalLabels(metric *prometheus.CounterVec, obj metav1.Object, fixedLabels map[string]string,
```

Comment:
I tried to make the logging of different types of metrics more dynamic, unfortunately the way `WithLabelValues`, `With`, `CurryWith`, `Observe` etc are implemented are different for histograms and summaries vs counters, and I couldn't find a generic interface to allow for it.

Reply:
Right, so this begs for an Interface that we define. And in fact, this may help us with the boggle I posed above wrt how consumers could discover which metrics are affected. To wit:

If we define an interface like

```go
type MetricWithOptionalLabels interface {
    CurryOptionalLabels(map[string]string)
}
```

...then we can define thin supertypes for the metrics we care about, e.g.

```go
type CounterMetricWithOptionalLabels struct {
    prometheus.CounterVec
}

// Assert we've implemented the method right
var _ MetricWithOptionalLabels = CounterMetricWithOptionalLabels{}

// CurryOptionalLabels implements MetricWithOptionalLabels
func (metric *CounterMetricWithOptionalLabels) CurryOptionalLabels(labels map[string]string) {
    // ... code here ...
}
```

The benefits are twofold:
1. When we declare the metrics in question, we declare them as their appropriate subtype, and then we can just call `CurryOptionalLabels()` on them.
2. Our docstring can say, "Affected metrics are instances of types that implement the MetricWithOptionalLabels interface," and consumers can go do some fancy F12ing in VSCode or whatever to figure out which ones those are.

Reply:
Cool. Assuming there's nowhere we need to curry without observing or vice versa, that's a cleaner model anyway.